### ISSUE TITLE
Mouse Selection on OSX failing after upgrading to 2018.10.27 (homebrew)

### ISSUE THREAD

**Issue by leeola** [mawww/kakoune#2651]
### Steps 

Scrolling up or down. Making mouse selections.

### Outcome 

Scrolling up seems to work, but scrolling down does not. No mouse selection works.

### Expected 

It should scroll down and select text.

OSX: Mojave
Kakoune: 2018-10-27

After some searching it sounds like this might be related to ncurses v5? I'm unsure how to tell what version of ncurses Kakoune is using. I've got v6 installed via brew, but maybe that's not enough.

Advice appreciated.

**Comment by Delapouite** [comment]
Hello. Which terminal emulator do you use on OSX?

**Comment by Screwtapello** [comment]
> Scrolling up seems to work, but scrolling down does not.

This is a classic symptom of the ncurses v5 API.

> No mouse selection works.

This is not. If Kakoune receives scroll-up events, which are delivered as mouse-clicks of button 64, it's very weird that it wouldn't be able to receive ordinary mouse-clicks.

Just to verify: if you execute the command `:echo -debug "ui options: %opt{ui_options}"`  and then check the debug buffer with `:buffer *debug*`, are there any option listed? Specifically, is the `ncurses_enable_mouse` option listed?

**Comment by lenormf** [comment]
https://github.com/mawww/kakoune/issues/1701#issuecomment-345416092

>One thing you can try is to play with set global ui_options ncurses_wheel_down_button=<number>

**Comment by lenormf** [comment]
Related #2642 #439 #1701

**Comment by leeola** [comment]
> Hello. Which terminal emulator do you use on OSX?

iTerm2

> Just to verify: if you execute the command :echo -debug "ui options: %opt{ui_options}" and then check the debug buffer with :buffer *debug*, are there any option listed? Specifically, is the ncurses_enable_mouse option listed?

Looks like no options listed:

```
ui options: 
```

To be clear, choosing a new cursor location with the mouse does work. However selecting a range of text just places the cursor at the sight of the initial mouse click. No selection is made while holding mouse1 and dragging.

**Comment by Screwtapello** [comment]
So that's just iTerm2, no tmux or screen or other intermediate layers?

What is `$TERM` set to in the shell where you start Kakoune?

**Comment by leeola** [comment]
Sorry, iTerm with tmux. Fwiw, scrolling / selection *did* work in the old version of Kakoune, and it still works of course in tmux.

```
lee ~> echo $TERM
screen-256color
```

**Comment by leeola** [comment]
Setting the following inside of Kakoune fixes scrolling down. After this change, scrolling up and down behave normally.

```
set global ui_options ncurses_wheel_down_button=64
```

Selection changes drastically however. It's a bit hard to describe, but it looks like when I click and drag to select a piece of text, Kakoune starts scrolling upward. This is new behavior with the above `ncurses_wheel_down_button` setting.

**Comment by Screwtapello** [comment]
Terminals generally support a number of different mouse event protocols, and the application (usually via a library like ncurses) asks for a particular protocol and parses the result.

Your `$TERM` mentions GNU Screen, and [the Screen documentation](https://www.gnu.org/software/screen/manual/html_node/Control-Sequences.html#Control-Sequences) says that Screen supports "X10 mouse tracking" (which only reports entire clicks, not individual presses and releases), and "VT200 mouse tracking" (which reports presses and releases, not movement). That might explain why dragging doesn't work.

Inside `tmux` try running:

    TERM=tmux-256color kak /path/to/some/file

...and see if that helps the mouse-dragging situation any.

You might also want to run `otool -L $(which kak)` to check which libraries it's using, and verify that one of them mentions "ncurses" and "6".

I guess also check the usual caveats: run `which kak` to verify that you're running the Kakoune binary you think you're running, and not some old leftover binary because the updated one was installed in a new location.

**Comment by joakin** [comment]
I can see the same thing, just installed kakoune with `brew install kakoune` and it seems to be using `libncurses.5.4` for me  on OSX 10.14.1. Maybe it is an issue with the homebrew formula.

```
> otool -L $(which kak)
/usr/local/bin/kak:
	/usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
```

**Comment by Screwtapello** [comment]
@joakin Can you confirm that you get the "can't scroll down" behaviour but *not* the "can't drag a selection" behaviour?

**Comment by joakin** [comment]
* I can scroll up (by scrolling down, like OSX does by default)
* scrolling down (by scrolling up) doesn't seem to work
* I can change the cursor position by clicking around
* I can't drag the mouse to select, the cursor stays where the mouse went down

**Comment by Screwtapello** [comment]
I poked around in the source-code for ncurses 5.4 (from 2005, no less!)
and it seems that while it *does* support drag events, it doesn't
support them in the default configuration. You might have some luck
if you:

    export TERM=xterm-1003

...before starting Kakoune, but Apple might not have installed that
experimental terminal definition, and even if it works that won't fix
the scrolling situation.

I suspect the solution is to ask Homebrew to build Kakoune against a
recent ncurses.

**Comment by joakin** [comment]
I'm opening an issue. Do you know what would be the required version for kakoune to work well? I'm not familiar with neither kakoune, ncurses or homebrew's internals, I'd appreciate some guidance 🙇

**Comment by joakin** [comment]
I've opened https://github.com/Homebrew/homebrew-core/issues/35514 to see if they can help. I'd appreciate any more specific pointers for the maintainers there.

**Comment by Screwtapello** [comment]
Looking at [the ncurses formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/ncurses.rb), it's using a modern (6.x) version of ncurses, and it's not passing the `--with-abi-version=5` argument which would make it backward-compatible with these bugs, so it should be fine.

**Comment by joakin** [comment]
I've opened a PR with the dependency, tested locally and it works great.

**Comment by mawww** [comment]
So I guess this can be closed now ?

**Comment by joakin** [comment]
Just tested. Seems fixed for me (OSX 10.14.1)

Commands I run:

```
=> brew update
=> brew reinstall kakoune
==> Reinstalling kakoune
==> Downloading https://homebrew.bintray.com/bottles/kakoune-2018.10.27_1.mojave.bottle.tar.gz
==> Downloading from https://akamai.bintray.com/e4/e48ab23c42154ca9ca49590973ddb9a42ee56e03e90a3638179f7b93fdb354d9?__gda__=exp=1546559852~hmac=7489f5e87eb12a07a1ebb710fdda05a3c2e158051a0937e96b7329b6c593a2b
######################################################################## 100.0%
==> Pouring kakoune-2018.10.27_1.mojave.bottle.tar.gz
🍺  /usr/local/Cellar/kakoune/2018.10.27_1: 135 files, 3.2MB
=> kak ~/tmp/log.txt # Test scrolling
=> otool -L $(which kak)
/usr/local/bin/kak:
	/usr/local/opt/ncurses/lib/libncursesw.6.dylib (compatibility version 6.0.0, current version 6.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
```

**Comment by mawww** [comment]
Perfect, thanks for the feedback.
